torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -324,127 +258,141 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
+                from a call to :meth:`state_dict`
         """
-        Returns:
-            the last known global optimizer state, which consist of a list of the shards.
 
-        .. warning:
-            If the state has not been consolidated, this returns a shard's worth, not the global state.
+        for i_param, (key, value) in enumerate(state_dict["state"].items()):
+            param = self.index_to_param[i_param]
 
-        .. warning:
-            Returning the global state is limited to the replica which was responsible for the consolidation.
-            The state may also not be up to date, depending on when `consolidate_state_dict` was last called.
-        """
+            # Populate the sharded optimizer state on the fly
+            if self.param_to_rank[param] != self.rank:
+                state_dict["state"][key] = None
 
-        if len(self._all_states) == 0:
-            logging.warning("Optimizer state has not been consolidated. Returning the local state")
-            logging.warning("Please call `consolidate_state_dict()` beforehand if you meant to save the global state")
-            state_dict = self.local_state_dict()
-            state_dict["local_state_dict"] = True
-            return state_dict
-
-        # Flatten the param_groups, save the partition which logs the rank <> shard correspondence
-        partition: List[Tuple[int, int]] = []
-        param_groups: List[Dict[Any, Any]] = []
-
-        start = 0
-        for i, s in enumerate(self._all_states):
-            param_groups.extend(s["param_groups"])
-            end = start + len(s["param_groups"])
-            partition.append((start, end))
-            start = end
+            if key in self.index_to_param:
+                param = self.index_to_param[i_param]
 
-        return {
-            "state": [s["state"] for s in self._all_states],
-            "param_groups": param_groups,
-            "partition": partition,
-            "local_state_dict": False,
-        }
+                # Only add this state to the sharded optimizer if it owns this param
+                for pg in self.optim.param_groups:
+                    if id(param) in map(lambda x: id(x), pg["params"]):
+                        self.optim.state[param] = _recursive_copy_to_device(
+                            value, non_blocking=True, device=param.device
+                        )
 
-    @staticmethod
-    def rank_local_state_dict(rank: int, state_dict: dict) -> dict:
-        """Returns the local_state_dict for a given rank.
+        super().load_state_dict(state_dict)
 
-        Arguments:
-            rank (int): rank to get local_state_dict for
-            state_dict (dict): global state_dict
-        """
-        param_groups = state_dict["param_groups"][state_dict["partition"][rank][0] : state_dict["partition"][rank][1]]
-        return {"state": state_dict["state"][rank], "param_groups": param_groups}
+        # Sync with the optimizer param groups
```

Comment:
sync all the param groups from the saved dict to the root param group here (the one publicly exposed), then propagate this to the shard optim